### PR TITLE
refactor(rubygems): Use `Result` type for caching control flow

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Replace ad-hoc result union type with dedicated `Result<T, E>` type
- Errors re-thrown are still intact in this PR

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/rubygems/index.ts [hunk 1]
```diff
@@ -41,13 +41,11 @@ export class RubyGemsDatasource extends Datasource {
     }
 
     try {
-      const cachedVersions = await this.versionsEndpointCache.getVersions(
-        registryUrl,
-        packageName
-      );
+      const { res: versionsResult } =
```
Comment (viceice) [inline lib/modules/datasource/rubygems/index.ts:44]: can we use the old variable name to have less unrelated changes?

File: lib/modules/datasource/rubygems/versions-endpoint-cache.ts [hunk 2]
```diff
@@ -140,28 +142,26 @@ export class VersionsEndpointCache {
    * At any given time, there should only be one request for a given registryUrl.
    */
   private async getCache(registryUrl: string): Promise<VersionsEndpointResult> {
-    const cacheKey = `rubygems-versions-cache:${registryUrl}`;
+    const oldResult = memCache.get(registryUrl);
```
Comment (viceice) [inline lib/modules/datasource/rubygems/versions-endpoint-cache.ts:145]: why changing the cache key?

File: lib/modules/datasource/rubygems/index.ts (hunk 1 above)
Comment (zharinov) [reply lib/modules/datasource/rubygems/index.ts:44]: It still will affect other lines because of the change of underlying structure. Because of this, I think it's better make different naming as well.

File: lib/modules/datasource/rubygems/versions-endpoint-cache.ts (hunk 2 above)
Comment (zharinov) [reply lib/modules/datasource/rubygems/versions-endpoint-cache.ts:145]: It isn't going to the long-term cache, this structure can only have one element. In further PRs, I'll make it a variable instead of `Set`, but for now it's enough to simplify this `cacheKey` thing. Otherwise, I would need to add `getCacheKey(registryUrl)` method instead, as it's being used in several places.

File: lib/modules/datasource/rubygems/versions-endpoint-cache.ts [hunk 3]
```diff
@@ -78,21 +73,28 @@ function parseFullBody(body: string): VersionsEndpointData {
...
+type VersionsEndpointResult = Result<VersionsEndpointData, 'unsupported-api'>;
 
 export const memCache = new Map<string, VersionsEndpointResult>();
 
+function cacheResult(
+  registryUrl: string,
+  result: VersionsEndpointResult
+): void {
+  const registryHostname = parseUrl(registryUrl)?.hostname;
+  if (registryHostname === 'rubygems.org') {
```
Comment (viceice) [inline lib/modules/datasource/rubygems/versions-endpoint-cache.ts:93]: why do we only cache rubygems.org in memory?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/modules/datasource/rubygems/versions-endpoint-cache.ts (hunk 3 above)
Comment (zharinov) [reply lib/modules/datasource/rubygems/versions-endpoint-cache.ts:93]: We don't want to cache private registries

**Review by rarkins — APPROVED** [review]
